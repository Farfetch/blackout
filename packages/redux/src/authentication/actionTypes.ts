/**
 * Action type dispatched when the login request fails.
 */
export const LOGIN_FAILURE = '@farfetch/blackout-redux/LOGIN_FAILURE';
/**
 * Action type dispatched when the login request starts.
 */
export const LOGIN_REQUEST = '@farfetch/blackout-redux/LOGIN_REQUEST';
/**
 * Action type dispatched when the login request succeeds.
 */
export const LOGIN_SUCCESS = '@farfetch/blackout-redux/LOGIN_SUCCESS';
/**
 * Action type dispatched when the client wants to reset the login request redux
 * data.
 */
export const LOGIN_RESET = '@farfetch/blackout-redux/LOGIN_RESET';

/**
 * Action type dispatched when the logout request fails.
 */
export const LOGOUT_FAILURE = '@farfetch/blackout-redux/LOGOUT_FAILURE';
/**
 * Action type dispatched when the logout request starts.
 */
export const LOGOUT_REQUEST = '@farfetch/blackout-redux/LOGOUT_REQUEST';
/**
 * Action type dispatched when the logout request succeeds.
 */
export const LOGOUT_SUCCESS = '@farfetch/blackout-redux/LOGOUT_SUCCESS';
/**
 * Action type dispatched when the client wants to reset the logout request redux
 * data.
 */
export const LOGOUT_RESET = '@farfetch/blackout-redux/LOGOUT_RESET';

/**
 * Action type dispatched when the change password request fails.
 */
export const PASSWORD_CHANGE_FAILURE =
  '@farfetch/blackout-redux/PASSWORD_CHANGE_FAILURE';
/**
 * Action type dispatched when the change password request starts.
 */
export const PASSWORD_CHANGE_REQUEST =
  '@farfetch/blackout-redux/PASSWORD_CHANGE_REQUEST';
/**
 * Action type dispatched when the change password request succeeds.
 */
export const PASSWORD_CHANGE_SUCCESS =
  '@farfetch/blackout-redux/PASSWORD_CHANGE_SUCCESS';
/**
 * Action type dispatched when the client wants to reset the change password
 * request redux data.
 */
export const PASSWORD_CHANGE_RESET =
  '@farfetch/blackout-redux/PASSWORD_CHANGE_RESET';

/**
 * Action type dispatched when the recover password request fails.
 */
export const PASSWORD_RECOVER_FAILURE =
  '@farfetch/blackout-redux/PASSWORD_RECOVER_FAILURE';
/**
 * Action type dispatched when the recover password request starts.
 */
export const PASSWORD_RECOVER_REQUEST =
  '@farfetch/blackout-redux/PASSWORD_RECOVER_REQUEST';
/**
 * Action type dispatched when the recover password request succeeds.
 */
export const PASSWORD_RECOVER_SUCCESS =
  '@farfetch/blackout-redux/PASSWORD_RECOVER_SUCCESS';
/**
 * Action type dispatched when the client wants to reset the recover password
 * request redux data.
 */
export const PASSWORD_RECOVER_RESET =
  '@farfetch/blackout-redux/PASSWORD_RECOVER_RESET';

/**
 * Action type dispatched when the reset password request fails.
 */
export const PASSWORD_RESET_FAILURE =
  '@farfetch/blackout-redux/PASSWORD_RESET_FAILURE';
/**
 * Action type dispatched when the reset password request starts.
 */
export const PASSWORD_RESET_REQUEST =
  '@farfetch/blackout-redux/PASSWORD_RESET_REQUEST';
/**
 * Action type dispatched when the reset password request succeeds.
 */
export const PASSWORD_RESET_SUCCESS =
  '@farfetch/blackout-redux/PASSWORD_RESET_SUCCESS';
/**
 * Action type dispatched when the client wants to reset the reset password request
 * redux data.
 */
export const PASSWORD_RESET_RESET =
  '@farfetch/blackout-redux/PASSWORD_RESET_RESET';

/**
 * Action type dispatched when the register request fails.
 */
export const REGISTER_FAILURE = '@farfetch/blackout-redux/REGISTER_FAILURE';
/**
 * Action type dispatched when the register request starts.
 */
export const REGISTER_REQUEST = '@farfetch/blackout-redux/REGISTER_REQUEST';
/**
 * Action type dispatched when the register request succeeds.
 */
export const REGISTER_SUCCESS = '@farfetch/blackout-redux/REGISTER_SUCCESS';
/**
 * Action type dispatched when the client wants to reset the register request redux
 * data.
 */
export const REGISTER_RESET = '@farfetch/blackout-redux/REGISTER_RESET';

/**
 * Action type dispatched when the validate email request fails.
 */
export const VALIDATE_EMAIL_FAILURE =
  '@farfetch/blackout-redux/VALIDATE_EMAIL_FAILURE';
/**
 * Action type dispatched when the validate email request starts.
 */
export const VALIDATE_EMAIL_REQUEST =
  '@farfetch/blackout-redux/VALIDATE_EMAIL_REQUEST';
/**
 * Action type dispatched when the validate email request succeeds.
 */
export const VALIDATE_EMAIL_SUCCESS =
  '@farfetch/blackout-redux/VALIDATE_EMAIL_SUCCESS';
/**
 * Action type dispatched when the client wants to reset the validate email request
 * redux data.
 */
export const VALIDATE_EMAIL_RESET =
  '@farfetch/blackout-redux/VALIDATE_EMAIL_RESET';

/**
 * Action type dispatched when the refresh email token request fails.
 */
export const REFRESH_EMAIL_TOKEN_FAILURE =
  '@farfetch/blackout-redux/REFRESH_EMAIL_TOKEN_FAILURE';
/**
 * Action type dispatched when the refresh email token request starts.
 */
export const REFRESH_EMAIL_TOKEN_REQUEST =
  '@farfetch/blackout-redux/REFRESH_EMAIL_TOKEN_REQUEST';
/**
 * Action type dispatched when the refresh email token request succeeds.
 */
export const REFRESH_EMAIL_TOKEN_SUCCESS =
  '@farfetch/blackout-redux/REFRESH_EMAIL_TOKEN_SUCCESS';
/**
 * Action type dispatched when the client wants to reset the validate email request
 * redux data.
 */
export const REFRESH_EMAIL_TOKEN_RESET =
  '@farfetch/blackout-redux/REFRESH_EMAIL_TOKEN_RESET';

/**
 * Action type dispatched when the delete user token request fails.
 */
export const DELETE_USER_TOKEN_FAILURE =
  '@farfetch/blackout-redux/DELETE_USER_TOKEN_FAILURE';
/**
 * Action type dispatched when the delete user token request starts.
 */
export const DELETE_USER_TOKEN_REQUEST =
  '@farfetch/blackout-redux/DELETE_USER_TOKEN_REQUEST';
/**
 * Action type dispatched when the delete user token request succeeds.
 */
export const DELETE_USER_TOKEN_SUCCESS =
  '@farfetch/blackout-redux/DELETE_USER_TOKEN_SUCCESS';
/**
 * Action type dispatched when the client wants to reset the delete user token
 * request redux data.
 */
export const DELETE_USER_TOKEN_RESET =
  '@farfetch/blackout-redux/DELETE_USER_TOKEN_RESET';

/**
 * Action type dispatched when the delete user impersonation request fails.
 */
export const DELETE_USER_IMPERSONATION_FAILURE =
  '@farfetch/blackout-redux/DELETE_USER_IMPERSONATION_FAILURE';
/**
 * Action type dispatched when the delete user impersonation request starts.
 */
export const DELETE_USER_IMPERSONATION_REQUEST =
  '@farfetch/blackout-redux/DELETE_USER_IMPERSONATION_REQUEST';
/**
 * Action type dispatched when the delete user impersonation request succeeds.
 */
export const DELETE_USER_IMPERSONATION_SUCCESS =
  '@farfetch/blackout-redux/DELETE_USER_IMPERSONATION_SUCCESS';
/**
 * Action type dispatched when the client wants to reset the delete user
 * impersonation request redux data.
 */
export const DELETE_USER_IMPERSONATION_RESET =
  '@farfetch/blackout-redux/DELETE_USER_IMPERSONATION_RESET';

/**
 * Action type dispatched when the create user impersonation request fails.
 */
export const CREATE_USER_IMPERSONATION_FAILURE =
  '@farfetch/blackout-redux/CREATE_USER_IMPERSONATION_FAILURE';
/**
 * Action type dispatched when the create user impersonation request starts.
 */
export const CREATE_USER_IMPERSONATION_REQUEST =
  '@farfetch/blackout-redux/CREATE_USER_IMPERSONATION_REQUEST';
/**
 * Action type dispatched when the create user impersonation request succeeds.
 */
export const CREATE_USER_IMPERSONATION_SUCCESS =
  '@farfetch/blackout-redux/CREATE_USER_IMPERSONATION_SUCCESS';
/**
 * Action type dispatched when the client wants to reset the create user
 * impersonation request redux data.
 */
export const CREATE_USER_IMPERSONATION_RESET =
  '@farfetch/blackout-redux/CREATE_USER_IMPERSONATION_RESET';

/**
 * Action type dispatched when the create user token request fails.
 */
export const CREATE_USER_TOKEN_FAILURE =
  '@farfetch/blackout-redux/CREATE_USER_TOKEN_FAILURE';
/**
 * Action type dispatched when the create user token request starts.
 */
export const CREATE_USER_TOKEN_REQUEST =
  '@farfetch/blackout-redux/CREATE_USER_TOKEN_REQUEST';
/**
 * Action type dispatched when the create user token request succeeds.
 */
export const CREATE_USER_TOKEN_SUCCESS =
  '@farfetch/blackout-redux/CREATE_USER_TOKEN_SUCCESS';
/**
 * Action type dispatched when the client wants to reset the create user token
 * request redux data.
 */
export const CREATE_USER_TOKEN_RESET =
  '@farfetch/blackout-redux/CREATE_USER_TOKEN_RESET';

/**
 * Action type dispatched when the create client credentials request fails.
 */
export const CREATE_CLIENT_CREDENTIALS_TOKEN_FAILURE =
  '@farfetch/blackout-redux/CREATE_CLIENT_CREDENTIALS_TOKEN_FAILURE';
/**
 * Action type dispatched when the create client credentials request starts.
 */
export const CREATE_CLIENT_CREDENTIALS_TOKEN_REQUEST =
  '@farfetch/blackout-redux/CREATE_CLIENT_CREDENTIALS_TOKEN_REQUEST';
/**
 * Action type dispatched when the create client credentials request succeeds.
 */
export const CREATE_CLIENT_CREDENTIALS_TOKEN_SUCCESS =
  '@farfetch/blackout-redux/CREATE_CLIENT_CREDENTIALS_TOKEN_SUCCESS';
/**
 * Action type dispatched when the client wants to reset the create client
 * credentials token request redux data.
 */
export const CREATE_CLIENT_CREDENTIALS_TOKEN_RESET =
  '@farfetch/blackout-redux/CREATE_CLIENT_CREDENTIALS_TOKEN_RESET';

/**
 * Action type dispatched when the refresh user token request fails.
 */
export const REFRESH_USER_TOKEN_FAILURE =
  '@farfetch/blackout-redux/REFRESH_USER_TOKEN_FAILURE';
/**
 * Action type dispatched when the refresh user token request starts.
 */
export const REFRESH_USER_TOKEN_REQUEST =
  '@farfetch/blackout-redux/REFRESH_USER_TOKEN_REQUEST';
/**
 * Action type dispatched when the refresh user token request succeeds.
 */
export const REFRESH_USER_TOKEN_SUCCESS =
  '@farfetch/blackout-redux/REFRESH_USER_TOKEN_SUCCESS';
/**
 * Action type dispatched when the client wants to reset the refresh user token
 * request redux data.
 */
export const REFRESH_USER_TOKEN_RESET =
  '@farfetch/blackout-redux/REFRESH_USER_TOKEN_RESET';

/**
 * Action type dispatched when the client wants to reset the token request redux
 * data.
 */
export const AUTHENTICATION_RESET =
  '@farfetch/blackout-redux/AUTHENTICATION_RESET';
