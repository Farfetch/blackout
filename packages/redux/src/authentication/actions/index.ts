/**
 * Authentication actions.
 *
 * @module authentication/actions
 * @category Authentication
 * @subcategory Actions
 */
export { default as changePassword } from './changePassword';
export { default as createClientCredentialsToken } from './createClientCredentialsToken';
export { default as createUserImpersonation } from './createUserImpersonation';
export { default as createUserToken } from './createUserToken';
export { default as login } from './login';
export { default as logout } from './logout';
export { default as recoverPassword } from './recoverPassword';
export { default as refreshEmailToken } from './refreshEmailToken';
export { default as refreshToken } from './refreshToken';
export { default as register } from './register';
export { default as removeUserImpersonation } from './removeUserImpersonation';
export { default as removeUserToken } from './removeUserToken';
export { default as reset } from './reset';
export { default as resetPassword } from './resetPassword';
export { default as validateEmail } from './validateEmail';
