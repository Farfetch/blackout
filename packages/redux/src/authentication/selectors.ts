/**
 * @module authentication/selectors
 * @category Authentication
 * @subcategory Selectors
 */

import {
  getChangePassword,
  getError,
  getIsLoading,
  getLogin,
  getLogout,
  getRecoverPassword,
  getRefreshEmailToken,
  getRegister,
  getResetPassword,
  getUserImpersonation,
  getUserToken,
  getValidateEmail,
} from './reducer';
import { getUser } from '../entities/selectors';

/**
 * Returns the status of the user authentication status.
 *
 * @function
 * @param {object} state - Application state.
 * @param {object} state.user - User entity.
 * @param {boolean} state.user.isGuest - User type.
 * @param {number} state.user.id - User id.
 * @returns {boolean} IsAuthenticated.
 */
export const isAuthenticated = (state: {
  user: { isGuest: boolean; id: number };
}): boolean =>
  Boolean(getUser(state) && !getUser(state).isGuest && getUser(state).id);

/**
 * Returns the error or loading status of each sub-area.
 */

/**
 * Returns the loading status for the login operation.
 *
 * @function
 *
 * @param {object} state - Application state.
 *
 * @returns {boolean} Login Loading status.
 */
export const isLoginLoading = (state: any): boolean =>
  getLogin(state.authentication).isLoading;

/**
 * Returns the error for the login operation.
 *
 * @function
 *
 * @param {object} state - Application state.
 *
 * @returns {object} Login operation error.
 */
export const getLoginError = (state: any): Record<string, unknown> =>
  getLogin(state.authentication).error;

/**
 * Returns the loading status for the logout operation.
 *
 * @function
 *
 * @param {object} state - Application state.
 *
 * @returns {boolean} Logout Loading status.
 */
export const isLogoutLoading = (state: any): boolean =>
  getLogout(state.authentication).isLoading;

/**
 * Returns the error for the logout operation.
 *
 * @function
 *
 * @param {object} state - Application state.
 *
 * @returns {object} Logout operation error.
 */
export const getLogoutError = (state: any): Record<string, unknown> =>
  getLogout(state.authentication).error;

/**
 * Returns the loading status for the change password operation.
 *
 * @function
 *
 * @param {object} state - Application state.
 *
 * @returns {boolean} Change password Loading status.
 */
export const isChangePasswordLoading = (state: any): boolean =>
  getChangePassword(state.authentication).isLoading;

/**
 * Returns the error for the change password operation.
 *
 * @function
 *
 * @param {object} state - Application state.
 *
 * @returns {object} Change password operation error.
 */
export const getChangePasswordError = (state: any): Record<string, unknown> =>
  getChangePassword(state.authentication).error;

/**
 * Returns the loading status for the recover password operation.
 *
 * @function
 *
 * @param {object} state - Application state.
 *
 * @returns {boolean} Recover password Loading status.
 */
export const isRecoverPasswordLoading = (state: any): boolean =>
  getRecoverPassword(state.authentication).isLoading;

/**
 * Returns the error for the recover password operation.
 *
 * @function
 *
 * @param {object} state - Application state.
 *
 * @returns {object} Recover password operation error.
 */
export const getRecoverPasswordError = (state: any): Record<string, unknown> =>
  getRecoverPassword(state.authentication).error;

/**
 * Returns the loading status for the reset password operation.
 *
 * @function
 *
 * @param {object} state - Application state.
 *
 * @returns {boolean} Reset password Loading status.
 */
export const isResetPasswordLoading = (state: any): boolean =>
  getResetPassword(state.authentication).isLoading;

/**
 * Returns the error for the reset password.
 *
 * @function
 *
 * @param {object} state - Application state.
 *
 * @returns {object} Reset password error.
 */
export const getResetPasswordError = (state: any): Record<string, unknown> =>
  getResetPassword(state.authentication).error;

/**
 * Returns the loading status for the register operation.
 *
 * @function
 *
 * @param {object} state - Application state.
 *
 * @returns {boolean} Register Loading status.
 */
export const isRegisterLoading = (state: any): boolean =>
  getRegister(state.authentication).isLoading;

/**
 * Returns the error for the register operation.
 *
 * @function
 *
 * @param {object} state - Application state.
 *
 * @returns {object} Register operation error.
 */
export const getRegisterError = (state: any): Record<string, unknown> =>
  getRegister(state.authentication).error;

/**
 * Returns the loading status for the validate email operation.
 *
 * @function
 *
 * @param {object} state - Application state.
 *
 * @returns {boolean} Validate email Loading status.
 */
export const isValidateEmailLoading = (state: any): boolean =>
  getValidateEmail(state.authentication).isLoading;

/**
 * Returns the error for the validate email operation.
 *
 * @function
 *
 * @param {object} state - Application state.
 *
 * @returns {object} Validate email error.
 */
export const getValidateEmailError = (state: any): Record<string, unknown> =>
  getValidateEmail(state.authentication).error;

/**
 * Returns the loading status for the refresh email token operation.
 *
 * @function
 *
 * @param {object} state - Application state.
 *
 * @returns {boolean} Refresh email token operation Loading status.
 */
export const isRefreshEmailTokenLoading = (state: any): boolean =>
  getRefreshEmailToken(state.authentication).isLoading;

/**
 * Returns the refresh email token error.
 *
 * @function
 *
 * @param {object} state - Application state.
 *
 * @returns {object} Refresh email token operation error.
 */
export const getRefreshEmailTokenError = (
  state: any,
): Record<string, unknown> => getRefreshEmailToken(state.authentication).error;

/**
 * Returns the loading status for the user token operation.
 *
 * @function
 *
 * @param {object} state - Application state.
 *
 * @returns {boolean} User token operation Loading status.
 */
export const isUserTokenLoading = (state: any): boolean =>
  getUserToken(state.authentication).isLoading;

/**
 * Returns the get user token error.
 *
 * @function
 *
 * @param {object} state - Application state.
 *
 * @returns {object} Get user token operation error.
 */
export const getUserTokenError = (state: any): Record<string, unknown> =>
  getUserToken(state.authentication).error;

/**
 * Returns the result for the get user token operation.
 *
 * @function
 *
 * @param {object} state - Application state.
 *
 * @returns {object} Get user token operation result.
 */
export const getUserTokenResult = (state: any): Record<string, unknown> =>
  getUserToken(state.authentication).result;

/**
 * Returns the loading status for the user impersonation operation.
 *
 * @function
 *
 * @param {object} state - Application state.
 *
 * @returns {boolean} User impersonation operation Loading status.
 */
export const isUserImpersonationLoading = (state: any): boolean =>
  getUserImpersonation(state.authentication).isLoading;

/**
 * Returns the get user impersonation error.
 *
 * @function
 *
 * @param {object} state - Application state.
 *
 * @returns {object} User impersonation operation error.
 */
export const getUserImpersonationError = (
  state: any,
): Record<string, unknown> => getUserImpersonation(state.authentication).error;

/**
 * Returns the result for the user impersonation operation.
 *
 * @function
 *
 * @param {object} state - Application state.
 *
 * @returns {object} User impersonation operation result.
 */
export const getUserImpersonationResult = (
  state: any,
): Record<string, unknown> => getUserImpersonation(state.authentication).result;

/**
 * Returns the loading  for the authentication.
 *
 * @function
 *
 * @param {object} state - Application state.
 *
 * @returns {boolean} Authentication Loading status.
 */
export const isAuthenticationLoading = (state: any): boolean =>
  getIsLoading(state.authentication);

/**
 * Returns the authentication error.
 *
 * @function
 *
 * @param {object} state - Application state.
 *
 * @returns {object} Authentication error.
 */
export const getAuthenticationError = (state: any): Record<string, unknown> =>
  getError(state.authentication);
