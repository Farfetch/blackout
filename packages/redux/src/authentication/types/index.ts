export * from './action';
export * from './state.types';
