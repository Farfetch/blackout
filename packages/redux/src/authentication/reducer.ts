import * as actionTypes from './actionTypes';
import { CombinedState, combineReducers } from 'redux';
import { reducerFactory } from '../helpers';
interface Action {
  type: string;
  payload?: {
    result?: any;
    error?: any;
  };
}

const INITIAL_STATE = {
  error: null,
  id: null,
  isLoading: false,
  login: {
    error: null,
    isLoading: false,
  },
  logout: {
    error: null,
    isLoading: false,
  },
  register: {
    error: null,
    isLoading: false,
  },
  changePassword: {
    error: null,
    isLoading: false,
  },
  recoverPassword: {
    error: null,
    isLoading: false,
  },
  resetPassword: {
    error: null,
    isLoading: false,
  },
  validateEmail: {
    error: null,
    isLoading: false,
  },
  refreshEmailToken: {
    error: null,
    isLoading: false,
  },
  userToken: {
    result: null,
    error: null,
    isLoading: false,
  },
  userImpersonation: {
    result: null,
    error: null,
    isLoading: false,
  },
};

const error = (state = INITIAL_STATE.error, action: Action) => {
  switch (action?.type) {
    case actionTypes.LOGIN_FAILURE:
    case actionTypes.LOGOUT_FAILURE:
    case actionTypes.REGISTER_FAILURE:
    case actionTypes.PASSWORD_CHANGE_FAILURE:
    case actionTypes.PASSWORD_RECOVER_FAILURE:
    case actionTypes.PASSWORD_RESET_FAILURE:
      return action?.payload?.error;
    case actionTypes.AUTHENTICATION_RESET:
    case actionTypes.LOGIN_REQUEST:
    case actionTypes.LOGOUT_REQUEST:
    case actionTypes.PASSWORD_CHANGE_REQUEST:
    case actionTypes.PASSWORD_RECOVER_REQUEST:
    case actionTypes.PASSWORD_RESET_REQUEST:
    case actionTypes.REGISTER_REQUEST:
      return INITIAL_STATE.error;
    default:
      return state;
  }
};

const id = (state = INITIAL_STATE.id, action: Action) => {
  switch (action?.type) {
    case actionTypes.LOGIN_SUCCESS:
    case actionTypes.REGISTER_SUCCESS:
      return action?.payload?.result;
    case actionTypes.AUTHENTICATION_RESET:
    case actionTypes.LOGOUT_SUCCESS:
      return INITIAL_STATE.id;
    default:
      return state;
  }
};

const isLoading = (state = INITIAL_STATE.isLoading, action: Action) => {
  switch (action?.type) {
    case actionTypes.LOGIN_REQUEST:
    case actionTypes.LOGOUT_REQUEST:
    case actionTypes.REGISTER_REQUEST:
    case actionTypes.PASSWORD_CHANGE_REQUEST:
    case actionTypes.PASSWORD_RECOVER_REQUEST:
    case actionTypes.PASSWORD_RESET_REQUEST:
      return true;
    case actionTypes.LOGIN_SUCCESS:
    case actionTypes.LOGOUT_SUCCESS:
    case actionTypes.REGISTER_SUCCESS:
    case actionTypes.PASSWORD_CHANGE_SUCCESS:
    case actionTypes.PASSWORD_RECOVER_SUCCESS:
    case actionTypes.PASSWORD_RESET_SUCCESS:
      return INITIAL_STATE.isLoading;
    default:
      return state;
  }
};

export const login = reducerFactory('LOGIN', INITIAL_STATE.login, actionTypes);
export const logout = reducerFactory(
  'LOGOUT',
  INITIAL_STATE.logout,
  actionTypes,
);
export const register = reducerFactory(
  'REGISTER',
  INITIAL_STATE.register,
  actionTypes,
);
export const changePassword = reducerFactory(
  'PASSWORD_CHANGE',
  INITIAL_STATE.changePassword,
  actionTypes,
);
export const resetPassword = reducerFactory(
  'PASSWORD_RESET',
  INITIAL_STATE.resetPassword,
  actionTypes,
);
export const recoverPassword = reducerFactory(
  'PASSWORD_RECOVER',
  INITIAL_STATE.recoverPassword,
  actionTypes,
);
export const validateEmail = reducerFactory(
  'VALIDATE_EMAIL',
  INITIAL_STATE.validateEmail,
  actionTypes,
);
export const refreshEmailToken = reducerFactory(
  'REFRESH_EMAIL_TOKEN',
  INITIAL_STATE.refreshEmailToken,
  actionTypes,
);
export const userToken = (
  state = INITIAL_STATE.userToken,
  action: Action,
): { result: any; error: any; isLoading: any } => {
  switch (action?.type) {
    case actionTypes.DELETE_USER_TOKEN_REQUEST:
    case actionTypes.CREATE_USER_TOKEN_REQUEST:
    case actionTypes.CREATE_CLIENT_CREDENTIALS_TOKEN_REQUEST:
    case actionTypes.REFRESH_USER_TOKEN_REQUEST:
      return {
        ...state,
        error: INITIAL_STATE.userToken.error,
        isLoading: true,
      };
    case actionTypes.DELETE_USER_TOKEN_SUCCESS:
      return {
        result: INITIAL_STATE.userToken.result,
        error: INITIAL_STATE.userToken.error,
        isLoading: false,
      };

    case actionTypes.CREATE_USER_TOKEN_SUCCESS:
    case actionTypes.CREATE_CLIENT_CREDENTIALS_TOKEN_SUCCESS:
    case actionTypes.REFRESH_USER_TOKEN_SUCCESS:
      return {
        result: action?.payload,
        error: INITIAL_STATE.userToken.error,
        isLoading: false,
      };

    case actionTypes.DELETE_USER_TOKEN_FAILURE:
    case actionTypes.CREATE_USER_TOKEN_FAILURE:
    case actionTypes.CREATE_CLIENT_CREDENTIALS_TOKEN_FAILURE:
    case actionTypes.REFRESH_USER_TOKEN_FAILURE:
      return {
        ...state,
        error: action?.payload?.error,
        isLoading: false,
      };

    default:
      return state;
  }
};
export const userImpersonation = (
  state = INITIAL_STATE.userImpersonation,
  action: Action,
): { result: any; error: any; isLoading: any } => {
  switch (action?.type) {
    case actionTypes.DELETE_USER_IMPERSONATION_REQUEST:
    case actionTypes.CREATE_USER_IMPERSONATION_REQUEST:
      return {
        ...state,
        error: INITIAL_STATE.userImpersonation.error,
        isLoading: true,
      };
    case actionTypes.CREATE_USER_IMPERSONATION_SUCCESS:
      return {
        result: action?.payload,
        error: INITIAL_STATE.userImpersonation.error,
        isLoading: false,
      };

    case actionTypes.DELETE_USER_IMPERSONATION_SUCCESS:
      return {
        result: INITIAL_STATE.userImpersonation.result,
        error: INITIAL_STATE.userImpersonation.error,
        isLoading: false,
      };

    case actionTypes.DELETE_USER_IMPERSONATION_FAILURE:
    case actionTypes.CREATE_USER_IMPERSONATION_FAILURE:
      return {
        ...state,
        error: action?.payload?.error,
        isLoading: false,
      };

    default:
      return state;
  }
};

/**
 * Export entities action mapper.
 */

export const entitiesMapper = {
  [actionTypes.LOGOUT_SUCCESS as typeof actionTypes.LOGOUT_SUCCESS]: (
    state: CombinedState<any>,
  ): CombinedState<any> => {
    return {
      ...state,
      user: {},
    };
  },
};

export const getError = (state: { error?: any }): any => state.error;
export const getIsLoading = (state: { isLoading: boolean }): boolean =>
  state.isLoading;
export const getId = (state: { id: number }): number => state.id;
export const getLogin = (state: { login: any }): any => state.login;
export const getLogout = (state: { logout: any }): any => state.logout;
export const getRegister = (state: { register: any }): any => state.register;
export const getChangePassword = (state: { changePassword: any }): any =>
  state.changePassword;
export const getResetPassword = (state: { resetPassword: any }): any =>
  state.resetPassword;
export const getRecoverPassword = (state: { recoverPassword: any }): any =>
  state.recoverPassword;
export const getValidateEmail = (state: { validateEmail: any }): any =>
  state.validateEmail;
export const getRefreshEmailToken = (state: { refreshEmailToken: any }): any =>
  state.refreshEmailToken;
export const getUserToken = (state: { userToken: any }): any => state.userToken;
export const getUserImpersonation = (state: { userImpersonation: any }): any =>
  state.userImpersonation;

/**
 * Reducer for authentication state.
 *
 * @param state  - Current redux state.
 * @param action - Action dispatched.
 *
 * @returns New state.
 */
export default combineReducers({
  error,
  id,
  isLoading,
  login,
  logout,
  register,
  changePassword,
  resetPassword,
  recoverPassword,
  validateEmail,
  refreshEmailToken,
  userToken,
  userImpersonation,
});
