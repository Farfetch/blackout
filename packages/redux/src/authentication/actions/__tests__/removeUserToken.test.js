import { deleteTokens } from '@farfetch/blackout-client/authentication';
import { mockStore } from '../../../../tests';
import { removeUserToken } from '..';
import find from 'lodash/find';
import reducer, { actionTypes } from '../..';

jest.mock('@farfetch/blackout-client/authentication', () => ({
  ...jest.requireActual('@farfetch/blackout-client/authentication'),
  deleteTokens: jest.fn(),
}));

const authenticationMockStore = (state = {}) =>
  mockStore({ authentication: reducer() }, state);

const expectedConfig = undefined;
let store;

describe('removeUserToken() action creator', () => {
  const userTokenId = 'dsjfn-dsfnksdj';

  beforeEach(() => {
    jest.clearAllMocks();
    store = authenticationMockStore();
  });

  it('should create the correct actions for when the delete user token procedure fails', async () => {
    const errorObject = {
      errorMessage: 'delete user token error',
      errorCode: 0,
      status: 400,
    };

    deleteTokens.mockRejectedValueOnce(errorObject);
    expect.assertions(4);

    try {
      await store.dispatch(removeUserToken(userTokenId));
    } catch (error) {
      expect(error).toBe(errorObject);
      expect(deleteTokens).toHaveBeenCalledTimes(1);
      expect(deleteTokens).toHaveBeenCalledWith(userTokenId, expectedConfig);
      expect(store.getActions()).toEqual(
        expect.arrayContaining([
          { type: actionTypes.DELETE_USER_TOKEN_REQUEST },
          {
            type: actionTypes.DELETE_USER_TOKEN_FAILURE,
            payload: { error: errorObject },
          },
        ]),
      );
    }
  });

  it('should create the correct actions for when the delete user token procedure is successful', async () => {
    const mockResponse = {
      status: 204,
    };

    deleteTokens.mockResolvedValueOnce(mockResponse);
    await store.dispatch(removeUserToken(userTokenId));

    const actionResults = store.getActions();

    expect(deleteTokens).toHaveBeenCalledTimes(1);
    expect(deleteTokens).toHaveBeenCalledWith(userTokenId, expectedConfig);

    expect(actionResults).toMatchObject([
      { type: actionTypes.DELETE_USER_TOKEN_REQUEST },
      {
        type: actionTypes.DELETE_USER_TOKEN_SUCCESS,
      },
    ]);
    expect(
      find(actionResults, {
        type: actionTypes.DELETE_USER_TOKEN_SUCCESS,
      }),
    ).toMatchSnapshot('delete user token success payload');
  });
});
