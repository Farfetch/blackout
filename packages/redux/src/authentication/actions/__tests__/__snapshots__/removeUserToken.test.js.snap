// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`removeUserToken() action creator should create the correct actions for when the delete user token procedure is successful: delete user token success payload 1`] = `
Object {
  "meta": Object {
    "userTokenId": "dsjfn-dsfnksdj",
  },
  "type": "@farfetch/blackout-redux/DELETE_USER_TOKEN_SUCCESS",
}
`;
